### PR TITLE
Add cause to Weather/ThunderChangeEvents

### LINKED ISSUE #3397: Add `Cause` to `WeatherChangeEvent` and `ThunderChangeEvent`
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
Create something similar to [LightningStrikeEvent.Cause](https://papermc.io/javadocs/paper/1.15/org/bukkit/event/weather/LightningStrikeEvent.Cause.html) for [WeatherChangeEvent](https://papermc.io/javadocs/paper/1.15/org/bukkit/event/weather/WeatherChangeEvent.html) and [ThunderChangeEvent](https://papermc.io/javadocs/paper/1.15/org/bukkit/event/weather/ThunderChangeEvent.html)

**Describe alternatives you've considered**
...

**Additional context**
...

### PR DESCRIPTION
**Pull request by Machine-Maker**
closes #3397

### CODE REVIEW

**Comment by Machine-Maker** [comment]
oh no you dont

**Comment by Machine-Maker** [comment]
`Rebased after the NMS repackage`

**Review by MiniDigger — APPROVED** [review]
lgtm, only concern I have with those setStorm methods without the new arg, shouldnt that just call the OG method, but with UNKOWN as cause?
Else when mojang adds new usages, we will not know about them.

**Comment by Machine-Maker** [comment]
@MiniDigger wouldn't we know about them? cause it wouldn't compile if its just missing an argument. Or maybe I am misunderstanding what you mean.

**Comment by MiniDigger** [comment]
no it would still call the old method, since that still exists with the same signature, right?

normally, when we add a new param to a method, we do
```
// Paper start
public void method(Old arg) {
	method(arg, DEFAULT);
}
public void method(Old arg, New arg) {
	// Paper end
	// old code
}
```

**Comment by Machine-Maker** [comment]
But the old method doesn’t exist with the same signature. I added an argument. So any new places would be missing that argument, and it wouldn’t compile. By adding a new method and leaving the old one, **then** you end up missing when mojang adds new calls to it without our reason.

**Comment by Machine-Maker** [comment]
yeah, idk what I was taking about, refactored (and rebased) this to make more sense. @MiniDigger

File: Spigot-Server-Patches/0741-Add-cause-to-Weather-ThunderChangeEvents.patch [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+-        this.worldDataServer.setStorm(false);
++        this.worldDataServer.setStorm(false, org.bukkit.event.weather.WeatherChangeEvent.Cause.COMMAND); // Paper - happens when weather cycle gamerule is set to false
+         // If we stop due to everyone sleeping we should reset the weather duration to some other random value.
+         // Not that everyone ever manages to get the whole server to sleep at the same time....
+         if (!this.worldDataServer.hasStorm()) {
+             this.worldDataServer.setWeatherDuration(0);
+         }
+         // CraftBukkit end
+-        this.worldDataServer.setThundering(false);
++        this.worldDataServer.setThundering(false, org.bukkit.event.weather.ThunderChangeEvent.Cause.COMMAND); // Paper - happens when weather cycle gamerule is set to false
```
Comment (kennytv) [inline Spigot-Server-Patches/0741-Add-cause-to-Weather-ThunderChangeEvents.patch]: No, this happens when players go to sleep to skip the night.

File: Spigot-Server-Patches/0741-Add-cause-to-Weather-ThunderChangeEvents.patch [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
++                this.worldDataServer.setStorm(flag2, org.bukkit.event.weather.WeatherChangeEvent.Cause.COMMAND); // Paper
+             }
+ 
+             this.lastThunderLevel = this.thunderLevel;
+@@ -885,14 +885,14 @@ public class WorldServer extends World implements GeneratorAccessSeed {
+ 
+     private void clearWeather() {
+         // CraftBukkit start
+-        this.worldDataServer.setStorm(false);
++        this.worldDataServer.setStorm(false, org.bukkit.event.weather.WeatherChangeEvent.Cause.COMMAND); // Paper - happens when weather cycle gamerule is set to false
```
Comment (kennytv) [inline Spigot-Server-Patches/0741-Add-cause-to-Weather-ThunderChangeEvents.patch]: No, this happens when players go to sleep to skip the night.

File: Spigot-Server-Patches/0741-Add-cause-to-Weather-ThunderChangeEvents.patch [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+ 
+     public BiomeBase getBiomeBySeed(int i, int j, int k) { return a(i, j, k); } // Paper - OBFHELPER
+@@ -482,8 +482,8 @@ public class WorldServer extends World implements GeneratorAccessSeed {
+                 this.worldDataServer.setThunderDuration(j);
+                 this.worldDataServer.setWeatherDuration(k);
+                 this.worldDataServer.setClearWeatherTime(i);
+-                this.worldDataServer.setThundering(flag1);
+-                this.worldDataServer.setStorm(flag2);
++                this.worldDataServer.setThundering(flag1, org.bukkit.event.weather.ThunderChangeEvent.Cause.NATURAL); // Paper
++                this.worldDataServer.setStorm(flag2, org.bukkit.event.weather.WeatherChangeEvent.Cause.COMMAND); // Paper
```
Comment (kennytv) [inline Spigot-Server-Patches/0741-Add-cause-to-Weather-ThunderChangeEvents.patch]: Why is the reason here `COMMAND`?

**Review by kennytv — CHANGES_REQUESTED** [review]
~~You missed a setStorm/setThunder call when a world is being loaded~~

**Comment by Machine-Maker** [comment]
mmkay, made those changes.

**Review by kennytv — APPROVED** [review]

**Pull request merged**